Test driver Integration

desaikd opened the pull request:
**Description:**

This PR is regarding changes for integrating ion-js with[ ion-test-driver](https://github.com/amzn/ion-test-driver).

**Completetd Tasks:**

- Make test-driver separate from ion-js
   - remove compareEquivs from `IonEventStream `
   - remove use of IonCliError and IonComparisonReport from `IonEventstream`.
   - remove use of equivs method from`iontests`
   - Placed comparisonResult in src for using it in compare methods.
   - modularize the test-driver code. 

- test-driver package settings
   - create package.json for test-driver.
   - create tsconfig for test-driver.
   - add grunt task to build test-driver.
   - add gitignore files for test-driver. 

**TO DO:**

- remove unnecessary import from `iontests`.

desaikd committed:
Implement process for ion-js cli

desaikd committed:
Revision 2 -  Implement process for ion-js-cli

desaikd committed:
embedded streams and event stream to ion stream conversion

desaikd committed:
added unknown text support for annotations

desaikd committed:
implement compare for ion-js-cli test-driver

desaikd committed:
fixed according to changes reuested

desaikd committed:
fixed changes requested and merge conflicts

desaikd committed:
remove duplicate write to file and remove unsused import

desaikd committed:
added yargs to dependencies and small spell fix

desaikd committed:
bug fixes for equivs comparison and ion tests

desaikd committed:
modified process for embedded streams

desaikd committed:
modified according to suggested changes

desaikd committed:
added comment for 'any' data type issue

desaikd committed:
test-driver integration changes

zslayton commented on .gitignore:
```diff
@@ -42,6 +44,7 @@ tramp
 # cask packages
 .cask/
 dist/
+test-driver/dist
```
Looks like this entry is already on line 3 above.

zslayton commented on .gitignore:
```diff
@@ -1,7 +1,9 @@
+test-driver/dist
 node_modules/
 src/.baseDir.ts
+test-driver/src/.baseDir.ts
```
What creates `.baseDir.ts`? If it's generated by some aspect of our tooling, do we want to ignore `**/.baseDir.ts` instead?

desaikd replied:
I think when compiling the files it gets created in the directory specified by the rootDirs option in tsconfig. but yes, can change to  **/.baseDir.ts.

zslayton commented on Gruntfile.js:
```diff
@@ -41,6 +42,9 @@ module.exports = function(grunt) {
         tsconfig: 'tsconfig.commonjs.json'
+      },
+      'test-driver' : {
+        tsconfig: './test-driver/tsconfig.test-driver.json'
```
Since this config files is no longer an extension of another one (and lives in a directory by itself), I think we can get away with naming it `test-driver/tsconfig.json`. Thoughts?

desaikd replied:
Yes. I think that makes it easier to understand.

zslayton commented on Gruntfile.js:
```diff
@@ -17,6 +17,7 @@ module.exports = function(grunt) {
   grunt.initConfig({
     pkg: grunt.file.readJSON('package.json'),
     clean: ['dist/',
+            'test-driver/dist/',
```
Does this belong in the top-level `clean` command rather than in a `test-driver:clean` or similar?

desaikd replied:
Yes, I can change clean to have two lists. One to have ion-js and test-driver, as the directories to be cleaned needs to be defined here with the clean option.

zslayton commented on Gruntfile.js:
```diff
@@ -185,7 +191,7 @@ module.exports = function(grunt) {
 
   grunt.registerTask('cleanup', 'Removes extraneous files from distributions', function() {
-    grunt.file.expand('dist/**/IonEvent*').forEach((file) => grunt.file.delete(file));
+    //grunt.file.expand('dist/**/test-driver/*.d.ts').forEach((file) => grunt.file.delete(file));
```
I think this comment can be removed?

desaikd replied:
I think I can remove this task it is not required anymore for any cleanup.

zslayton commented on src/Ion.ts:
```diff
@@ -139,3 +139,8 @@ export {dom};
 // Re-export dom convenience methods for easy access via 'ion'
 export {load, loadAll} from "./dom";
+
+//Ion-event exports and comparison Result export
```
Nit: Should `Ion-event` be `IonEvent` or `ion-event`? (The package or the class?) I'm unsure based on the `export`s that follow.

desaikd replied:
I think it should rather be `events` as it has both `IonEvent` and `IonEventStream` exports here.

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -1,59 +1,73 @@
+        writer.writeSymbol("$ion_event_stream");
         for (let i: number = 0; i < this.eventStream.length; i++) {
-            this.eventStream[i].write(writer);
+                this.eventStream[i].write(writer);
```
Nit: spacing is a little off here. `prettier` will probably end up changing all of this to 2-space indentation though, so 🤷‍♂️

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -1,59 +1,73 @@
 
 export class IonEventStream {
-    private eventStream: IonEvent[];
+    eventStream: IonEvent[];
```
Two thoughts here: 
1. Could `eventStream` and `isEventStream` be `private`, or is there a reason we need them to be publicly visible? I notice there's a `getEvents()` function defined below.
2. Because `eventStream` is in a class called `IonEventStream`, I expected it to also be of type `IonEventStream`. Could we name it something less recursive like `events`? (This is obviously subjective, so feel free to push back.)

desaikd replied:
1. I had to make `eventStream` and `isEventStream` public to use it in test-driver code so that I could separate out the test-driver code from `IonEventStream`.

2. I think that makes sense will change it to events.

zslayton replied:
> I had to make eventStream and isEventStream public to use it in test-driver code so that I could separate out the test-driver code from IonEventStream.

I think you can use `eventStream` already by calling the public `getEvents()` method to access it. Similarly, you could add a public `isEventStream()` method to access `isEventStream`.

desaikd replied:
But in the case of embedded stream I want to modify the `eventStream` elements I can use a method called `setEvents()` and make `eventStream` private again and similarly create `isEventStream()`.  Does this sound good?

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -93,27 +107,39 @@ export class IonEventStream {
         let actualIndex: number = 0;
         let expectedIndex: number = 0;
+        if(this.eventStream.length != expected.eventStream.length) {
+            return new ComparisonResult(ComparisonResultType.ERROR, "Both event streams of different length");
```
```suggestion
            return new ComparisonResult(ComparisonResultType.ERROR, "The event streams have different lengths");
```

Also: why is this an `ERROR` rather than just `NOT_EQUAL`?

desaikd replied:
Yes, I think that is correct it should be NOT_EQUAL.

zslayton commented on src/ComparisonResult.ts:
```diff
@@ -0,0 +1,37 @@
+
+/** Comparison result types for the comparison report */
+export enum ComparisonResultType {
+    EQUAL = "EQUAL",
+    NOT_EQUAL = "NOT_EQUAL",
+    ERROR = "ERROR"
```
Could you add some comments explaining what each of these enum variants mean? (In particular, what does `ERROR` mean and when is it different from `NOT_EQUAL`?)

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -146,49 +183,61 @@ export class IonEventStream {
+                tid = this.reader.next();
             }
-            tid = this.reader.next();
+        } catch (er) {
```
Nit: `er` isn't a commonly used abbreviation for "error." How about `err` or just `error`?

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -146,49 +183,61 @@ export class IonEventStream {
             }
-            tid = this.reader.next();
+        } catch (er) {
+            throw new Error("READ " + er.message);
```
The `"READ"` prefix might help us track down this line of code, but it isn't very friendly to a non-developer who's just running the tests. Can we provide more context for someone who's just seeing the error message in their console?

desaikd replied:
Again to separate out the test-driver code from IonEventStream I had to use this prefix which I can use in the test-driver code for differentiation of Errors. I will add a comment for this.

zslayton replied:
Ah, I didn't realize the `READ $MESSAGE` format was part of the `test-driver` spec.

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -379,7 +428,25 @@ export class IonEventStream {
             this.reader.stepIn();
             for (let tid; tid = this.reader.next();) {
-                annotations.push(this.reader.stringValue()!);
+                if(tid == IonTypes.STRUCT) {
```
Here and elsewhere: TS style guides encourage using the "threequals" (`===`) operator by default and only falling back to `==` in situations where that behavior is required.

We can either try to fix this manually in this PR or fix amzn/ion-js#631.

desaikd replied:
I think doing it in  amzn#631 sounds good.

zslayton commented on test-driver/src/ComparisonContext.ts:
```diff
@@ -0,0 +1,102 @@
+import {IonEvent, IonEventFactory, IonEventType} from "ion-js";
+
+/**
+ * ComparisonContext to create a structure for lhs and rhs streams
```
Here and elsewhere: for classes whose structure/purpose is part of a spec, I think it would be helpful to provide a link to the full definition. That makes it less necessary to fully explain everything (including phrases like "lhs" and "rhs") in the class-level comments. 

I found [this](https://github.com/amzn/ion-test-driver#comparisonreport), but maybe there's something even better to point to?

zslayton commented on test-driver/src/Compare.ts:
```diff
@@ -0,0 +1,215 @@
+        }
+    }
+
+    compareFiles(ionOutputWriter: Writer, args: IonCompareArgs): void {
```
Throughout the new `test-driver` code: please add some doc comments to the publicly visible methods. (Probably not necessary for simple `constructor`s or accessor methods.)

desaikd replied:
Sure.

zslayton commented on test-driver/tsconfig.test-driver.json:
```diff
@@ -0,0 +1,27 @@
+// Base Typscript Compiler Configuration
+// Targets ES6 Module system and ES6 feature set.
```
I think this targets `"module": "commonjs"`?

zslayton commented on test-driver/tsconfig.test-driver.json:
```diff
@@ -0,0 +1,27 @@
+    "outDir": "dist",
+    "experimentalDecorators": true
+  }
+}
```
I think the file can just be named `test-driver/tsconfig.json` since there aren't any other `tsconfig`s in the same directory.

desaikd committed:
requested changes for test-driver integration

desaikd committed:
added more chnages for test-driver integration

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -229,39 +230,40 @@ export class IonEventStream {
-            throw new Error("READ " + er.message);
+        } catch (error) {
+            // This Error will be used by the test-driver to differentiate errors using error types(prefix).
+            throw new EventStreamError("READ" , error.message , this.events.length);
```
This is a good change -- I like that `READ` is an independently addressable/testable part of the `Error` now.

Instead of the string `"READ"` or `"WRITE"`, could this just be an instance of `ErrorType`?

desaikd replied:
You mean creating two different Error classes for READ and WRITE?

desaikd replied:
Oh I understand. But ErrorType is in test-driver code and wanted to keep them separate hence I used the string. I could declare Error Type in EventStreamError if we want to.

zslayton replied:
> I could declare Error Type in EventStreamError if we want to.

This sounds okay to me. This is a pretty minor concern anyway, we could merge the change even if we don't want to move `ErrorType`.

desaikd committed:
added scripts to package.json to build test-driver

desaikd committed:
added code to write eventstream before throwing erroring

desaikd committed:
changes to check null before comparison for eventstreams

desaikd committed:
bug fix for comparison

desaikd committed:
added more bug fixes to event streams

desaikd committed:
added Error handling for comparison

desaikd committed:
bug fixes for compare-equivs and event stream error

desaikd commented:
**Completed Tasks:**
1. added scripts in package.json to run and build test-driver from root directory ion-js. (0dc2a74)
2. added code to write event stream with available events before throwing error. (31544b3)
3. check null for event streams in all (basic, equivs and non-equivs) comparisons. (b02596e)
4. added parseFieldName with support of symbolTokens. (616b677)
5. changes to give correct event index when writing comparison report and Cli Errors. (8eb0f62)

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -482,4 +484,29 @@ export class IonEventStream {
         return this.reader.value();
     }
+
+    private parseFieldName(): string | null {
```
This function name doesn't quite communicate what it's doing; I would expect it to call `this.reader.fieldName()` and attempt to interpret that value somehow. Instead, it traverses a struct that contains either a `text` or `importLocation` field (but not both) and returns the value associated with whichever one it finds.

Could you rename it and add a comment giving an overview of its responsibilities?

desaikd replied:
So for the ion-test-driver the field_name for events is a symbol token. that is why I am parsing it here. Let me add a comment for that.

zslayton replied:
I think what's making this confusing to me is that "symbol token" is usually an in-memory representation of an `(ID, text, import location)` triple, and doesn't typically refer to a particular serialized value. In this case, "field name" isn't a string, it's a struct representing a serialized symbol token, which isn't obvious.

Can you do one or more of the following?:
* Link to some documentation showing what the serialized value it's expecting looks like
* Provide a sample value in the comment so readers can see an example of what it's expecting
* Rename the method to something more specific. Maybe `resolveFieldNameFromSerializedSymbolToken`? (Unfortunately verbose, but this is a strange case IMO.)

zslayton commented on src/events/IonEventStream.ts:
```diff
@@ -482,4 +484,29 @@ export class IonEventStream {
+        }
+        else if(this.reader.fieldName() == "importLocation" && type == IonTypes.INT) {
+            let symtab = defaultLocalSymbolTable();
+            let symbol = symtab.getSymbolText(this.reader.numberValue()!)
```
I'm surprised to see that `type` is `INT` but we're interpreting the value as a `symbol` ID. Is this part of the `test-driver` spec?

zslayton commented on test-driver/src/Compare.ts:
```diff
@@ -175,20 +178,26 @@ export class Compare {
                         && eventResult.result == ComparisonResultType.NOT_EQUAL) {
                         if(comparisonReport) {
-                            comparisonReport.writeComparisonReport(ComparisonResultType.NOT_EQUAL, eventResult.message, i + 1, j + 1);
+                            // set event-index as the index of container + (i/j) representing the event-index inside container
+                            // + 1 to skip the first event (the container event)
```
This is a really helpful comment.

zslayton commented on test-driver/src/ComparisonContext.ts:
```diff
@@ -50,16 +50,31 @@ export class ComparisonContext {
+            if(EventStreamError.eventstream) {
+                events.push(...EventStreamError.eventstream);
+            }
+            if(EventStreamError.type === "READ") {
```
If we can't use the `ErrorType` enum for `READ` and `WRITE`, we should still try to have constants for those strings that we can refer to. I know this is a "where should it live?" question, and you'll have more context on this than I do, but we should try to find a home for it.

desaikd replied:
Yes, I will add constants for these strings.

zslayton commented on test-driver/src/ComparisonContext.ts:
```diff
@@ -50,16 +50,31 @@ export class ComparisonContext {
+            this.eventStream = eventStream;
+        } catch (EventStreamError) {
+            if(EventStreamError.eventstream) {
+                events.push(...EventStreamError.eventstream);
```
In the past we've been burned by using `array.push(...someValues)` because it can [run into configured JS VM limits on the number of parameters a function can accept](https://github.com/amzn/ion-js/issues/601).

This is just an FYI; I don't think it's likely to be an issue for `EventStream` since our test cases aren't typically very long.

desaikd committed:
modified according to the changes requested

desaikd committed:
added changes for field name parsing

desaikd committed:
resolved git conflicts and ran prettier

zslayton commented on src/dom/README.md:
```diff
@@ -188,44 +199,46 @@ assert.equal(i, 7);
+| `ion.dom.Blob`        |   `Uint8Array`   |
+| `ion.dom.List`        |     `Array`      |
+| `ion.dom.SExpression` |     `Array`      |
+| `ion.dom.Struct`      |     `Object`     |
```
Wow, I didn't expect it to tidy up markdown tables. Nice! 👌
